roachtest: move decl_schema_change/job-compat to new framework

annrpom opened the pull request:
### roachtest: move decl_schema_change/job-compat to new framework

This patch updates the
`declarative_schema_changer/job-compatibility-mixed-version-V222-V231` roachtest to use the new mixed version testing framework - which in turn provides testing enhancements.

Informs: #110536
Epic: None

Release note: None

---

### roachtest: update version of job compat mixed-version roachtest

This patch updates the list of supported DDLs/the job compat roachtest
to declarative_schema_changer/job-compatibility-mixed-version-V231-V232.

Fixes: https://github.com/cockroachdb/cockroach/issues/110536
Epic: None

Release note: None

annrpom committed:
roachtest: move decl_schema_change/job-compat to new framework

This patch updates the
`declarative_schema_changer/job-compatibility-mixed-version-V222-V231`
roachtest to use the new mixed version testing framework - which in turn
provides testing enhancements.

Informs: #110536
Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114043)
<!-- Reviewable:end -->

annrpom commented:
CI lint issues - this is mostly good, I just need to make sure I didnt miss any DDLs and that the DDLs I am using (order, syntax) are good. going to use mixed-version backup's enable/disable job functions & leave the ones in backup.go for Renato (Renato, if you would like for me to delete these, feel free to lmk). 

Opening this for any nits that I can fix/any general questions, comments, concerns for what I have so far

rafiss reviewed (changes requested):
looks like it's on the right track!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-:-NjLdt-iD34D3SjZFMwe:b-ncmgav)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @DarrylWong, and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/backup.go` line 291 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjLdprDEEHGhv41e6K7:-NjLdprEDjXMqQmFc_Kw:b-hiuv6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d498aba280f06f90a4dc16cd8deddf6fe9a193a6/pkg/cmd/roachtest/tests/backup.go#L291)):*
> ```Go
> // registry from adopting a job.
> //
> // TODO(renato): remove this duplicated function once
> ```

based on this TODO, it looks like there might be another function we should use
___
*[`pkg/cmd/roachtest/tests/backup.go` line 355 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjLdhXuAxWldnctF7cC:-NjLdhXuAxWldnctF7cD:b-hiuv6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d498aba280f06f90a4dc16cd8deddf6fe9a193a6/pkg/cmd/roachtest/tests/backup.go#L355)):*
> ```Go
> // registry from adopting a job.
> //
> // TODO(renato): remove this duplicated function once
> ```

based on this TODO, it looks like there might be another function we should use
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjLbxgy3Cai-Pcb-9YO:-NjLbxgy3Cai-Pcb-9YP:b322v7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d498aba280f06f90a4dc16cd8deddf6fe9a193a6/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L34)):*
> ```Go
> 	// supported in the "previous" major release.
> 	r.Add(registry.TestSpec{
> 		Name:             "declarative_schema_changer/job-compatibility-mixed-version-V222-V231",
> ```

nit: we should rename this, since now we want to test `V231-V232`
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjLcIkyDWLcqpIy5oCJ:-NjLcIkyDWLcqpIy5oCK:b-895cxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d498aba280f06f90a4dc16cd8deddf6fe9a193a6/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L91)):*
> ```Go
> }
> 
> // executeSupportedDDLs tests all DDLs supported in V22_2 and V23_1.
> ```

i think as a separate commit in this same PR, we should change this function so it tests v23_1 <-> v23_2 compatibility.
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjLcurv6qOupcE89IJy:-NjLcurv6qOupcE89IJz:b-xdbmf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d498aba280f06f90a4dc16cd8deddf6fe9a193a6/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L168)):*
> ```Go
> ) {
> 	mvt := mixedversion.NewTest(
> 		ctx, t, t.L(), c, c.All(),
> ```

nit: this test specifically is testing compat between the last version and this one, so we should add a `mixedversion.NumUpgrades(1)` option


<!-- Sent from Reviewable.io -->

Xiang-Gu commented on pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go:
```diff
@@ -71,105 +78,120 @@ func planAndRunSchemaChange(
 	node option.NodeListOption,
 	schemaChangeStmt string,
-) {
+) error {
```
curious: Why do you add a `error` as the return? It seems it's not used anyway.

annrpom replied:
this gets bubbled up to our executeSupportedDDLs function; that error is then handled by the new mixed-version framework (since we call executeSupportedDDLs in `InMixedVersion`): https://github.com/cockroachdb/cockroach/blob/fbd1bb8b3699a999a14144229428683fd7e26fe7/pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go#L161-L169

Xiang-Gu commented on pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go:
```diff
@@ -12,56 +12,63 @@ package tests
+	// This test requires us to come back and change the to-be-tests stmts to be those
+	// supported in the "previous" major release.
+	r.Add(registry.TestSpec{
+		Name:             "declarative_schema_changer/job-compatibility-mixed-version-V231-V232",
```
I think in commit 1, this should be `V222-V231`?

annrpom replied:
yes, done - ty

Xiang-Gu commented on pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go:
```diff
@@ -71,105 +78,120 @@ func planAndRunSchemaChange(
+		// Disable job adoption for all nodes of the set we are testing [upgradedNodes, oldNodes] so that the
+		// other respective set can adopt the job (ex. for upgradedNodes, we want the oldNodes to be able to adopt
+		// these jobs).
+		if err := testUtils.disableJobAdoption(ctx, t.L(), nodes); err != nil {
```
great that we have this now!

Xiang-Gu commented on pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go:
```diff
@@ -109,25 +109,56 @@ func executeSupportedDDLs(
+		}
+
+		// DDLs supported since V23_1.
+		v231DDls := []string{
```
I think we also supported the following in V23_1 in DSC. Can you also include them here?
- `ADD FOREIGN KEY`
- `ADD FOREIGN KEY NOT VALID`
- `ADD UNIQUE WITHOUT INDEX`
- `ADD UNIQUE WITHOUT INDEX NOT VALID`
- `VALIDATE CONSTRAINT`
- `DROP CONSTRAINT`

Xiang-Gu reviewed (commented):
Good work! I've left a few comments and thanks for doing this!

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114043)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @DarrylWong, @renatolabs, and @Xiang-Gu)

___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 128 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114043#gh-1399456369:-NjsCa_NA848HgivXXop:bvtgdas) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6b6dcac7ede4e47fd082fa486720dd3c757aadc/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L128)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

I think we also supported the following in V23_1 in DSC. Can you also include them here?
- `ADD FOREIGN KEY`
- `ADD FOREIGN KEY NOT VALID`
- `ADD UNIQUE WITHOUT INDEX`
- `ADD UNIQUE WITHOUT INDEX NOT VALID`
- `VALIDATE CONSTRAINT`
- `DROP CONSTRAINT`
</blockquote></details>

do we need to use the `sql.schema.force_declarative_statements` cluster setting to make these use DSC when we want it?


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
Thanks for updating another test! :tada: 

A question about the test itself: I'm confused about why we have a specific upgrade in the test name ("V231-V232") but the test itself is running DDLs that are "supported since 22.2". Does it matter, then, what upgrade is being run, as long as it's at least V22_2?

IMO, having the version in the test name is confusing also because it might not reflect reality. For example, master will soon become 24.1, at which point we are not testing V231-V232 anymore. Should we skip the test at that point? From the test name, I'd say yes, but now that I took a closer look at the code, the answer is probably no (?).

I'm aware this PR is just updating an existing test, so these questions are optional and just for general consideration.

> Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-:-NjsxgSMDNxrxIxzhW27:b-n3ovem)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @DarrylWong, and @Xiang-Gu)

___
*[`pkg/cmd/roachtest/tests/backup.go` line 370 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-Njsl1VI68v5_sL6pamY:-Njsl1VI68v5_sL6pamZ:bdimv8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/013c7e6703244f541bece6a1037ccad080eb7616/pkg/cmd/roachtest/tests/backup.go#L370)):*
> ```Go
> 		require.NoError(t, err)
> 	}
> }
> ```

Thank you! :fire: 
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjslK_WBOJSJRwI4J_9:-NjslK_WBOJSJRwI4J_A:bv4px2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L31)):*
> to-be-tests stmts

Is this a typo? Not sure what this means.
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 54 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjsmPcb9Gm4Jzp123DB:-NjsmPcb9Gm4Jzp123DC:b-d82btv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L54)):*
> ```Go
> 		_, err := db.ExecContext(ctx, query, args...)
> 		return err
> 	}
> ```

Can we change `runQuery` to be:

```go
runQuery := func(query string, args ...interface{}) error {
    return h.Exec(r, query, args...)
}
```
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-Njsmg1F3gFCXMVHce9Q:-Njsmg1F3gFCXMVHce9R:b-dpz95w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L71)):*
> ```Go
> 		return err
> 	}
> 	return nil
> ```

This should be `h.Exec(r, "ALTER RANGE ...")`.
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 99 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjsqwBM45t_5O8r2aVY:-NjsqwBN55PpQ9u9kVFt:b-py9jhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L99)):*
> ```Go
> 		var nodes option.NodeListOption
> 		if testingUpgradedNodes {
> 			nodes = helper.Context().ToVersionNodes
> ```

This branch has drifted a bit from master. The functions you'll want to use for the purposes of this test are: `helper.Context.NodesInPreviousVersion()` and `helper.Context.NodesInNextVersion()`.

Also, something to be aware of is that we are not guaranteed to be in _mixed-binary_ state when this function is called. For example, it might be called when all nodes are running the master binary, but the cluster version is still fixed at `V23_1`. In this case, `NodesInPreviousVersion` will return an empty list. You can detect whether this is the case by using `helper.Context.MixedBinary()`.

I'll let your team decide what to do in this situation. A common approach is to just pick a random node in this case (example below). I think this is what should happen here as well since testing this scenario is important too.

https://github.com/cockroachdb/cockroach/blob/b847659cf51f39a0c5e03c3d946ecd475c501885/pkg/cmd/roachtest/tests/secondary_indexes.go#L67-L72
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 102 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjssOt5EGiMkhz377_f:-NjssOt5EGiMkhz377_g:b-1yiqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L102)):*
> ```Go
> 		}
> 		nodes = helper.Context().FromVersionNodes
> 		testUtils, err := newCommonTestUtils(ctx, t, c, nodes)
> ```

I believe it is currently technically harmless for the use of `testUtils` in this test, but that last argument (`nodes`) is supposed to encode the set of cockroach nodes in the cluster. You can use `helper.Context.CockroachNodes`.
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-Njst4lbBAyA-zA9baQ6:-Njst4lbBAyA-zA9baQ7:bnlan65) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L113)):*
> ```Go
> 		}
> 		node := option.NodeListOption{helper.RandomNode(r, nodes)}
> 		planAndRunSchemaChange(ctx, t, c, node, `COMMENT ON DATABASE testdb IS 'this is a database comment'`)
> ```

I don't think this `planAndRunSchemaChange` function is adding much -- IIUC, it just logs the statement and executes it. 

I suggest we drop that function and just have a list of statements, iterate over item and run them with `h.ExecWithGateway(rng, nodes, stmt)`.
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 143 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjstlMt-bEnBt4iCv6x:-NjstlMt-bEnBt4iCv6y:b2co78z) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L143)):*
> ```Go
> 
> 	// Set up the testing state (e.g. create a few databases and tables) and always use declarative schema
> 	// changer on all nodes. Queries run in this function should be idempotent.
> ```

Doesn't this function need to be called more than just `OnStartup`?  Some of the tables/sequences are dropped during the test.
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-Njstwf0BZ50KFUt0Nso:-Njstwf1A2sDIJV15ayM:b-19atck) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L158)):*
> ```Go
> 		CREATE SEQUENCE IF NOT EXISTS testdb.testsc.s;
> 		CREATE VIEW IF NOT EXISTS testdb.testsc.v AS (SELECT i*2 FROM testdb.testsc.t);
>     CREATE OR REPLACE FUNCTION fn(a INT) RETURNS INT AS 'SELECT a*a' LANGUAGE SQL;
> ```

Nit: Indentation on this last statement is off.
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 160 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-Njsu0nM9nD4BBsIdq9P:-Njsu0nM9nD4BBsIdq9Q:bs5z28t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L160)):*
> ```Go
>     CREATE OR REPLACE FUNCTION fn(a INT) RETURNS INT AS 'SELECT a*a' LANGUAGE SQL;
> `
> 		if _, err := db.ExecContext(ctx, setUpQuery); err != nil {
> ```

No need to create a new connection pool (`db` above) for this. We can use `h.Exec(r, setupQuery)`.
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 169 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjsuBS1-s1pLsKtJI2B:-NjsuBS1-s1pLsKtJI2C:brh7utu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L169)):*
> ```Go
> 		// buried by the fallback.
> 		for _, node := range c.All() {
> 			db := c.Conn(ctx, t.L(), node)
> ```

This is creating a new connection pool for every node, and not closing it (you'll see the related "leaked goroutines" list when you run this test on a gceworker).

I suggest using `h.Exec(r, "SET ...")`.
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 181 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjsuR2m2zFa4oLg_6Q9:-NjsuR2m2zFa4oLg_6QA:b-gb3b0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L181)):*
> ```Go
> 	mvt.OnStartup("set short job interval", setShortJobIntervalsStep)
> 	mvt.OnStartup("set short gcttl", setShortGCTTLInSystemZoneConfig)
> 	mvt.OnStartup("", testSetupResetStep)
> ```

Can we give this step a name?
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-Njsvh2g9Zsnhv5lX3Ib:-Njsvh2h9TYR3tH9jZpE:bvnknwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L194)):*
> ```Go
> 	// Testing that declarative schema change jobs created by nodes running older binary version can be adopted and
> 	// finished by nodes running newer binary versions.
> 	mvt.InMixedVersion("run test step on non upgraded node", executeSupportedDDLs(c, t, false /* testingUpgradedNodes */))
> ```

I believe we won't be able to have these steps as separate `InMixedVersion` functions. This means that they might possibly run in parallel, leading to a state where we disable job adoption on all nodes.


<!-- Sent from Reviewable.io -->

annrpom commented:
> Does it matter, then, what upgrade is being run, as long as it's at least V22_2?

I think the upgrade path here will be solely dependent on which branch we are on (as in - I don't believe we want to have a test plan where the upgrade starts at V22_2 on this branch since a group of the DDL statements that will be executed are only enabled on V23_1); thus, we will have to update this test whenever master changes to a new version

I guess it won't necessarily hurt if we execute these statements on V22_2 since I believe it will just test job compatibility in the legacy schema changer, but that sounds like it would miss the point of this test (?)

Maybe to make this point more evident, I can delete the batch of V22_2 statements on this branch (then the only place where this group will occur is on the backport)

annrpom committed:
roachtest: update version of job compat mixed-version roachtest

This patch updates the list of supported DDLs/the job compat roachtest
to declarative_schema_changer/job-compatibility-mixed-version-V231-V232.

Fixes: #110536
Epic: None

Release note: None

annrpom reviewed (commented):
> master will soon become 24.1, at which point we are not testing V231-V232 anymore

ah also yes - i'll make sure to update this to be V232-V241 accordingly once that changes

also, i just realized i have not published these comments :/

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-:-NkGEiiB1eUOl71WHOfl:bke7wwe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rafiss, @renatolabs, and @Xiang-Gu)

___
*[`pkg/cmd/roachtest/tests/backup.go` line 291 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjLdprDEEHGhv41e6K7:-NjePQPqD1NsqbYCYrS0:bfxy5ow) ([raw file](https://github.com/cockroachdb/cockroach/blob/d498aba280f06f90a4dc16cd8deddf6fe9a193a6/pkg/cmd/roachtest/tests/backup.go#L291)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

based on this TODO, it looks like there might be another function we should use
</blockquote></details>

yes - it looks like these aren't used anymore by anything @renatolabs let me know if you would like for me to keep these in
___
*[`pkg/cmd/roachtest/tests/backup.go` line 355 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjLdhXuAxWldnctF7cC:-NjePZMyAZyN05CQPjuc:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d498aba280f06f90a4dc16cd8deddf6fe9a193a6/pkg/cmd/roachtest/tests/backup.go#L355)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

based on this TODO, it looks like there might be another function we should use
</blockquote></details>

done
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjLbxgy3Cai-Pcb-9YO:-NjePhF-6aiKwzluKSWW:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d498aba280f06f90a4dc16cd8deddf6fe9a193a6/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L34)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: we should rename this, since now we want to test `V231-V232`
</blockquote></details>

done
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjLcIkyDWLcqpIy5oCJ:-NjhLqZU3Az0NQ5bXPdj:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d498aba280f06f90a4dc16cd8deddf6fe9a193a6/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L91)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think as a separate commit in this same PR, we should change this function so it tests v23_1 <-> v23_2 compatibility.
</blockquote></details>

done
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjLcurv6qOupcE89IJy:-NjePnwr0lpV3f-Lg750:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d498aba280f06f90a4dc16cd8deddf6fe9a193a6/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L168)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: this test specifically is testing compat between the last version and this one, so we should add a `mixedversion.NumUpgrades(1)` option
</blockquote></details>

done
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjslK_WBOJSJRwI4J_9:-NkGIuc749VvzHPaUrnr:b-pkvv5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L31)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

> to-be-tests stmts

Is this a typo? Not sure what this means.
</blockquote></details>

I reworded it - to me, this emphasized the fact that we will have to perform some maintenance every time master changes
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 54 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjsmPcb9Gm4Jzp123DB:-NkGJQAiBjhBdIGDkl1q:b-byp2e7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L54)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Can we change `runQuery` to be:

```go
runQuery := func(query string, args ...interface{}) error {
    return h.Exec(r, query, args...)
}
```
</blockquote></details>

indeed we can - done
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-Njsmg1F3gFCXMVHce9Q:-NkGMJ4_6bwbzkzjsMtr:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L71)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

This should be `h.Exec(r, "ALTER RANGE ...")`.
</blockquote></details>

done
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 99 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjsqwBM45t_5O8r2aVY:-NkRPCL2DdASRl9g-XWY:b-ghbw5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L99)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

This branch has drifted a bit from master. The functions you'll want to use for the purposes of this test are: `helper.Context.NodesInPreviousVersion()` and `helper.Context.NodesInNextVersion()`.

Also, something to be aware of is that we are not guaranteed to be in _mixed-binary_ state when this function is called. For example, it might be called when all nodes are running the master binary, but the cluster version is still fixed at `V23_1`. In this case, `NodesInPreviousVersion` will return an empty list. You can detect whether this is the case by using `helper.Context.MixedBinary()`.

I'll let your team decide what to do in this situation. A common approach is to just pick a random node in this case (example below). I think this is what should happen here as well since testing this scenario is important too.

https://github.com/cockroachdb/cockroach/blob/b847659cf51f39a0c5e03c3d946ecd475c501885/pkg/cmd/roachtest/tests/secondary_indexes.go#L67-L72
</blockquote></details>

ah good point - done
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 102 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjssOt5EGiMkhz377_f:-NkRZ78xCUdd5K8FR04a:b-9b5if7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L102)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I believe it is currently technically harmless for the use of `testUtils` in this test, but that last argument (`nodes`) is supposed to encode the set of cockroach nodes in the cluster. You can use `helper.Context.CockroachNodes`.
</blockquote></details>

done - thank you for catching this
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-Njst4lbBAyA-zA9baQ6:-NkRZDT4AraDsMdtJTyH:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L113)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I don't think this `planAndRunSchemaChange` function is adding much -- IIUC, it just logs the statement and executes it. 

I suggest we drop that function and just have a list of statements, iterate over item and run them with `h.ExecWithGateway(rng, nodes, stmt)`.
</blockquote></details>

done
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 169 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NjsuBS1-s1pLsKtJI2B:-NkW1qWy3dQu3Fz0KJ-3:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cdbc9792e1bab4d3cb2df6fcba7ca52d42ec73/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L169)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

This is creating a new connection pool for every node, and not closing it (you'll see the related "leaked goroutines" list when you run this test on a gceworker).

I suggest using `h.Exec(r, "SET ...")`.
</blockquote></details>

done
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 128 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114043#gh-1399456369:-NkGO0a-3TD9l5ERWi5k:bk9vq0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6b6dcac7ede4e47fd082fa486720dd3c757aadc/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L128)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do we need to use the `sql.schema.force_declarative_statements` cluster setting to make these use DSC when we want it?
</blockquote></details>

https://cockroachlabs.slack.com/archives/C04N0AS14CT/p1700596759258929 there was a discussion here about this question, but I clarified with Faizan and this is only needed for statements are not "on" (CREATE [SCHEMA, SEQUENCE])

the group of statements we are testing are all on by default - i don't think we need to do this for this PR


<!-- Sent from Reviewable.io -->

renatolabs reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Minor comments, this looks great! :clap: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-:-NlPN64kDYV7vAVWivWN:b-cpg9m2)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom, @DarrylWong, @rafiss, and @Xiang-Gu)

___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 75 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NlPKv_Y-LMQOPG2egm5:-NlPKv_Y-LMQOPG2egm6:b-fpxwew) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f73530c5ff0fab9854869442d0e005d3cada605/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L75)):*
> ```Go
> 	testingUpgradedNodes bool,
> ) error {
> 	nodes := c.All().RandNode()
> ```

I suggest we use `helper.RandomNode` instead to make sure we always pick the same node when running with the test with a specific seed (e.g., trying to reproduce a failure). `c.All().RandNode()` uses the global RNG, which is outside of our control.
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 186 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NlPMyp8C_Rhx1eJnLxQ:-NlPMyp8C_Rhx1eJnLxR:bz9wxu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f73530c5ff0fab9854869442d0e005d3cada605/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L186)):*
> ```Go
> 	mvt.OnStartup("set short gcttl", setShortGCTTLInSystemZoneConfig)
> 
> 	mvt.InMixedVersion("run test step", func(ctx context.Context, l *logger.Logger, r *rand.Rand, helper *mixedversion.Helper) error {
> ```

Should we run this function after the upgrade finishes too?


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114043)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DarrylWong, @rafiss, @renatolabs, and @Xiang-Gu)

___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 75 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NlPKv_Y-LMQOPG2egm5:-NlTjDHP879xs-wdmHLT:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f73530c5ff0fab9854869442d0e005d3cada605/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L75)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I suggest we use `helper.RandomNode` instead to make sure we always pick the same node when running with the test with a specific seed (e.g., trying to reproduce a failure). `c.All().RandNode()` uses the global RNG, which is outside of our control.
</blockquote></details>

done
___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 186 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-NlPMyp8C_Rhx1eJnLxQ:-NlTjIcp4OdO2pIdS3Gu:bxoawhj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f73530c5ff0fab9854869442d0e005d3cada605/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L186)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Should we run this function after the upgrade finishes too?
</blockquote></details>

it won't hurt if we do - done


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114043)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @annrpom, @DarrylWong, @renatolabs, and @Xiang-Gu)

___
*[`pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go` line 128 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114043#gh-1399456369:-NlRdpJL9G3F7oqsJchG:bpwlqym) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6b6dcac7ede4e47fd082fa486720dd3c757aadc/pkg/cmd/roachtest/tests/mixed_version_job_compatibility_in_declarative_schema_changer.go#L128)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

https://cockroachlabs.slack.com/archives/C04N0AS14CT/p1700596759258929 there was a discussion here about this question, but I clarified with Faizan and this is only needed for statements are not "on" (CREATE [SCHEMA, SEQUENCE])

the group of statements we are testing are all on by default - i don't think we need to do this for this PR
</blockquote></details>

i see - but i think that maybe we should test CREATE SCHEMA/SEQUENCE in this test if possible. since those were added during 23.2.


<!-- Sent from Reviewable.io -->

annrpom commented:
> since those were added during 23.2

I could be wrong, but I understood that the tests will test that the jobs created by statements implemented in the earlier testing version (so 23.1 in this case) could be picked up by newer versions (23.2) and vice versa for forwards and backwards compatibility

adding statements in 23.2 would mean that we would expect a job created by the dsc to be picked up by an earlier node version (and vice versa) - is that true? cc: @Xiang-Gu

rafiss reviewed (approved):
ah i see, yes i think that's right. this looks good from my end! just check if @renatolabs has any other comments

> Reviewed 1 of 2 files at r5, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-:-NlU1bhxAiBnozjybhW6:b-6fubc6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @annrpom, @DarrylWong, @renatolabs, and @Xiang-Gu)



<!-- Sent from Reviewable.io -->

renatolabs reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114043#-:-NlUTMWbAc1svLMrDyib:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom, @DarrylWong, @rafiss, and @Xiang-Gu)



<!-- Sent from Reviewable.io -->

annrpom commented:
TFTRs! ('-')7

bors r=rafiss, renatolabs

annrpom commented:
bors r-

annrpom commented:
needed to rebase

TFTRs! ('-')7

bors r=rafiss, renatolabs

annrpom commented:
blathers backport 23.2
